### DISCUSSION TITLE
Oidc-client: Custom dynamic headers

### DISCUSSION THREAD

**Discussion by rkraneis** [quarkusio/quarkus#22256] [Q&A] [answered]
We want to interface with an API using OpenID connect and on top of that the upcoming HTTP Message Signing standard. This requires us to modify the headers of the token requests based on the body contents before those token requests are sent out via Vert.x. As of now, we did not find a solution short of reimplementing OidcClient (to at least keep the token cashing infrastructure). Is there a better way (maybe vert.x equivalent of MessageWriterInterceptor)?

**Comment by sberyozkin** [comment, marked as answer]
@rkraneis Can you please clarify, do you refer to `quarkus-oidc-client` ?
In this case you can use the dynamic client creation feature, see: https://quarkus.io/guides/security-openid-connect-client#use-oidcclients.

So you can inject `OidcClients`, create `OidcClientConfig` dynamically (it allows to set the custom header) and create the client.
Does it help ?

**Reply by rkraneis** [reply]
Hi @sberyozkin, yes, this is about quarkus-oidc-client. The problem here is that I'd need to access the actual request body created by oidc-client (our counterpart requires POST to get a token, not GET) to calculate the value for the digest header - hence my mention of JAX RS MessageWriterInterceptor where I do exactly that for RestEasy. Unfortunately, the OIDC implementation is using Vert.x under the hood, and I cannot find an accessible integration point :(

**Reply by sberyozkin** [reply]
Hi @rkraneis 

`OidcClientImpl` does `client.postAbs(tokenRequestUri);`. Where do you see it doing `GET` ?

**Reply by rkraneis** [reply]
It is indeed always POST with secrets in the body or POST with secrets in the basic auth header. GET doesn't make sense.

**Reply by rkraneis** [reply]
But in the end that's tangential. “Relevant headers“ of every request to our counterpart must be signed; “Relevant headers” includes Content-Length and Digest (sha512sum of the body) when a body is present.

**Reply by rkraneis** [reply]
As I'm using RestEasy Client I can just use a MessageWriterInterceptor to accomplish this. But Ideally I'd like to use OIDC for automatic token management also ...

**Reply by sberyozkin** [reply]
I see, so when you said `our counterpart requires POST to get a token, not GET` I was not sure what you meant.

@cescoffier Hi Clement - can Vert.x Client be customized with the client filters ?

**Reply by rkraneis** [reply]
[HTTP Message Signatures][1] is the draft in questiion, @sberyozkin.

[1]: https://datatracker.ietf.org/doc/draft-ietf-httpbis-message-signatures/

**Reply by sberyozkin** [reply]
Sure, I'm aware of it, thanks. But the problem is how to intercept the Vert.x client request

**Reply by cescoffier** [reply]
@sberyozkin no, you cannot intercept the vertx http client traffic. You will need to do that directly, by wrapping the client.

**Reply by rkraneis** [reply]
Hmm, OK. Then I need to have a look at how OidcClientImpl is actually created in OidcClientRecorder and provide my own. Looked a bit overwhelming at first glance.

**Reply by sberyozkin** [reply]
@rkraneis I've looked at the draft, I'm not sure it is worth complicating it and the `OidcClients` idea mentioned earlier should work, unless I've missed something.  Do you need to set `SignatueInput` and `Signature` headers ? If yes, then unless you use a `private_key_jwt` client authentication method (which would produce a different data on every call), you can pre-calculate these header values, `OidcClient` will produce the same data on every call

**Reply by rkraneis** [reply]
@sberyozkin, yes, those are the headers. I need to check our counterparties documentation again, but I believe some non-static elements were strictly required. I will get back to you tomorrow.

**Reply by sberyozkin** [reply]
If the idea is viable in principle, then you can even set those values as static configuration properties (in `oidc-client.headers`)

**Reply by rkraneis** [reply]
That may undermine the intended security a bit too much, as I would have to fix the supposedly random nonce to a static value and put the required expires parameter far into the future. But with the known OIDC Json payload I may still set these values correctly if I can create a dynamic configuration source for OidcClientConfig.headers ...

**Reply by rkraneis** [reply]
Alas, this doesn't see to be possible. Most of the OidcClientConfig doesn't seem to be interceptable
```
mvn clean test | grep quarkus.oidc-client
2021-12-17 12:23:07,076 DEBUG [io.sma.config] (main) SRCFG01002: The config %test.quarkus.oidc-client-filter.register-filter was not found
2021-12-17 12:23:07,076 DEBUG [io.sma.config] (main) SRCFG01001: The config quarkus.oidc-client-filter.register-filter was loaded from default values with the value false
2021-12-17 12:23:07,078 DEBUG [io.sma.config] (main) SRCFG01002: The config %test.quarkus.oidc-client.enabled was not found
2021-12-17 12:23:07,078 DEBUG [io.sma.config] (main) SRCFG01002: The config quarkus.oidc-client.enabled was not found
2021-12-17 12:23:07,087 DEBUG [io.sma.config] (main) SRCFG01002: The config %test.quarkus.oidc-client-filter.client-name was not found
2021-12-17 12:23:07,087 DEBUG [io.sma.config] (main) SRCFG01002: The config quarkus.oidc-client-filter.client-name was not found
2021-12-17 12:23:07,171 DEBUG [io.sma.config] (main) SRCFG01001: The config %test.quarkus.oidc-client.early-tokens-acquisition was loaded from YamlConfigSource[source=file:/home/rene/Projects/upvest-integration/common/target/classes/application.yml] with the value false
2021-12-17 12:23:07,172 DEBUG [io.sma.config] (main) SRCFG01001: The config quarkus.oidc-client.early-tokens-acquisition was loaded from default values with the value true
2021-12-17 12:23:08,443 DEBUG [io.sma.config] (Lambda Thread (TEST)) SRCFG01002: The config %test.quarkus.oidc-client-filter.client-name was not found
2021-12-17 12:23:08,443 DEBUG [io.sma.config] (Lambda Thread (TEST)) SRCFG01002: The config quarkus.oidc-client-filter.client-name was not found
2021-12-17 12:23:08,443 DEBUG [io.sma.config] (Lambda Thread (TEST)) SRCFG01001: The config %test.quarkus.oidc-client.early-tokens-acquisition was loaded from YamlConfigSource[source=file:/home/rene/Projects/upvest-integration/common/target/classes/application.yml] with the value false
2021-12-17 12:23:08,443 DEBUG [io.sma.config] (Lambda Thread (TEST)) SRCFG01001: The config quarkus.oidc-client.early-tokens-acquisition was loaded from default values with the value true
```
I used the example LoggingConfigSourceInterceptor
```
@Priority(Priorities.LIBRARY + 200)
public class LoggingConfigSourceInterceptor implements ConfigSourceInterceptor {
  private static final long serialVersionUID = 367246512037404779L;

  @Override
  public ConfigValue getValue(final ConfigSourceInterceptorContext context, final String name) {
    ConfigValue configValue = doLocked(() -> context.proceed(name));
    if (configValue != null) {
      ConfigLogging.log.lookup(configValue.getName(), configValue.getLocation(), configValue.getValue());
    } else {
      ConfigLogging.log.notFound(name);
    }
    return configValue;
  }
}
```

**Reply by rkraneis** [reply]
Using a `ConfigSource` shows that the values are loaded once at runtime. With that, I can at least have the calculation of Signature and SignatureInput close to the remaining config and vary the values once per deployment. Not ideal, but workable for the moment.

**Reply by sberyozkin** [reply]
@rkraneis 

> That may undermine the intended security a bit too much, as I would have to fix the supposedly random nonce to a static value and put the required expires parameter far into the future. But with the known OIDC Json payload I may still set these values correctly if I can create a dynamic configuration source for OidcClientConfig.headers ...

and 

> Alas, this doesn't see to be possible. Most of the OidcClientConfig doesn't seem to be interceptable

Then as I said earlier, use the dynamic OidcClient creation

```
@GET
    public String getResponse() {
        OidcClientConfig cfg = new OidcClientConfig();
        cfg.setId("myclient");
        cfg.setAuthServerUrl("http://localhost:8081/auth/realms/quarkus/");
        cfg.setClientId("quarkus");
        cfg.getCredentials().setSecret("secret");
        
        // calculate the headers now, with a new nonce every time

        Uni<OidcClient> client = clients.newClient(config);
        // use this client to get the token
    }
```

If you also disable the discovery and only point to the token endpoint directly with the config  then the cost of creating a new client per every call will be close to 0...

**Reply by rkraneis** [reply]
@sberyozkin, thanks for the additional hint. Aren't acquired tokens bound to the current client instance? That's at least the impression I get from TokensHelper. If I create a fresh client for every intercepted request (I use ReastEassy to access the API), I will have a call to the token endpoint before every call also, right?

**Reply by rkraneis** [reply]
OK, after reading the code more carefully, I see that I just need to keep an instance of TokenHelper around. Thank you, that should solve the issue.

**Reply by sberyozkin** [reply]
@rkraneis Np, thanks, hope it will work well. If you'd like please create an enhancement request to support a better pluggability for OidcClient, I won't be able to prioritize in the short term but may be worth tracking such an enhancement request, thanks

**Reply by rkraneis** [reply]
Will, do. Unfortunately, improving OidcClient pluggability showed to not be the only missing piece in the puzzle. There's just no way within JAX-RS to modify headers based on the body after the registered MessageBodyWriter (e.g. Jackson) has done its part. So I'm currently serializing the payload twice. I'll have a look at RestEasy reactive client.
